File: Difficulty/Easy/[202] Happy Number/solution.cpp
/*
 * @lc app=leetcode id=202 lang=cpp
 *
 * [202] Happy Number
 */

// @lc code=start
class Solution {
public:
    bool isHappy(int n) {
        int slow = n, fast = n;
        do{
            slow = squareSum(slow);
            fast = squareSum(fast);
            fast = squareSum(fast);
        }while(slow != fast);

        return slow == 1;
    }

    int squareSum(int n) { 
        int res = 0;
        for(; n; n /= 10) {
            int num = n % 10;
            res += num * num;
        }
        return res;
    }
};
// @lc code=end

File: Difficulty/Easy/[202] Happy Number/solution.py
class Solution:
    def isHappy(self, n: int) -> bool:
        fast, slow, flag = n, n, True
        while flag or slow != fast:
            slow = self.squareSum(slow)
            fast = self.squareSum(fast)
            fast = self.squareSum(fast)
            flag = False
        return slow == 1

    def squareSum(self, n):
        res = 0
        while n:
            t = n % 10;
            res += t * t;
            n //= 10;
        return res;